### PR TITLE
Docker development environment

### PR DESCRIPTION
**Pull request by zuzust**
@andreslucena, something is wrong with Redis and Capybara configuration. Need to be checked out to make it work with Docker Compose.

For the sake of simplicity, here you are the docker commands needed to run development and test environments:

**Development**
```bash
$ docker-compose build
$ docker-compose run participa rake db:setup
$ docker-compose up -d
$ docker-compose exec participa mailcatcher --ip=0.0.0.0
$ docker-compose run -e "QUEUE=*" participa rake environment resque:work (???)
```

**Test**
```bash
$ docker-compose up -d
$ docker-compose run -e "RAILS_ENV=test" participa rake db:create db:migrate
$ docker-compose run -e "RAILS_ENV=test" --entrypoint="xvfb-run" participa bundle exec rake test
```

And here you are the key fragments of:

**database.yml**
```yaml
default: &default
  adapter: postgresql
  encoding: unicode
  pool: 5
  timeout: 5000
  username: postgres
  host: postgres
```

**secrets.yml**
```yaml
  features:
    allow_inscription: true
    blog: false
    collaborations: true
    collaborations_redsys: false
    elections: true
    microcredits: false
    notifications: false
    openid: true
    participation_teams: false
    proposals: false
    use_resque: false
    verification_presencial: true
    verification_sms: false
```

### CODE REVIEW

**Comment by andreslucena** [comment]
Regarding the Redis error, it seems like it's an issue with redis configuration, by default it's looking for localhost but with docker it isn't working there. We need a to tell it where to look it, like this: 

> 
> $ cat config/initializers/redis.rb
> 
> rails_env = ENV['RAILS_ENV'] || 'development'
> 
> if rails_env == 'development'
>   $redis = Redis.new(:host => 'redis')
> else
>   $redis = Redis.new
> end
> 
> Resque.redis = $redis
>

**Comment by zuzust** [comment]
@andreslucena You are right. I have added redis and mailserver config files to fix linking to Redis and Mailcatcher containers.

I have fixed the Capybara issue as well, but it seems there's something wrong with xvfb-run that prevents seeing the tests output from outside the container. Planning to try the Poltergeist + PhantomJS way.

By the way, when you are ok with the PR ping me and I will squash the commits.

**Comment by deivid-rodriguez** [comment]
:+1: to trying the Poltergeist +  PhantomJS way. Since phantomjs is natively headless, "the xvfb-run hack" shouldn't be needed.

**Comment by zuzust** [comment]
Here you are, a fully functional docker development environment. Replaced capybara-webkit for Poltergeist+PhantomJS successfully, although there's no Poltergeist alternative to capybara-webkit's `block_unknown_urls`.

**Comment by deivid-rodriguez** [comment]
@zuzust Great job!

There an alternative to `block_unknown_url`'s, see https://github.com/teampoltergeist/poltergeist#url-blacklisting--whitelisting. If this case, I think we want to blacklist everything.

**Comment by zuzust** [comment]
@deivid-rodriguez I read about it, but no examples of blacklisting everything :) Any ideas?

**Comment by deivid-rodriguez** [comment]
I normally keep a specific list of whitelisted urls, makes me aware of how much external JS I'm depending on. But according to [this comment](https://github.com/teampoltergeist/poltergeist/pull/701#issuecomment-157531436) and the following one, I think setting a whitelist to localhost would block everything else.

**Comment by zuzust** [comment]
@deivid-rodriguez Poltergeist whitelisting done! Check it out, and ping me if everything is ok to squash commits in this branch.

**Comment by deivid-rodriguez** [comment]
@zuzust Cool, I'm currently trying it :) I'll ping you back!

**Comment by deivid-rodriguez** [comment]
When running

```
$ docker-compose run participa rake db:setup
```
I get

```
Your bundle is locked to rake (11.1.2), but that version could not be found in any of the sources listed in your Gemfile. If you haven't changed sources, that means the author of rake (11.1.2) has removed it. You'll need to update your bundle to a different version of rake (11.1.2) that hasn't been removed in order to install.
```

I have to manually run 

```
docker-compose run participa bundle
```

before that to make the error go away.

This is because when running `bundle install --jobs 5 --retry 3` during `docker-compose build`, gems are installed globally to root's `GEM_PATH`. But when running `docker-compose run participa rake db:setup`, it's picking my sticky bundler config in `.bundle/config` which contains `BUNDLE_PATH: "vendor/bundle"`. So it can't find the right gems because they're not (yet) installed.

If this setup expects gems to live in `vendor/bundle` (I think this is the case), doing

```diff
diff --git a/config/docker/Dockerfile b/config/docker/Dockerfile
index 50d9e9c..0b1c690 100644
--- a/config/docker/Dockerfile
+++ b/config/docker/Dockerfile
@@ -62,7 +62,7 @@ ENV APP_HOME ${APP_HOME:-/app}
 WORKDIR $APP_HOME
 
 COPY Gemfile Gemfile.lock ./
-RUN bundle install --jobs 5 --retry 3
+RUN bundle install --path vendor/bundle --jobs 5 --retry 3
 COPY . ./
 
 RUN chown -R participa:participa ./
```
shoud fix this.

**Comment by zuzust** [comment]
Same output running `$ docker-compose run participa bundle exec rake db:setup`? Keep in mind that I've changed Dockerfile and now there's no implicit `bundle exec` as entrypoint.

**Comment by deivid-rodriguez** [comment]
> Same output running $ docker-compose run participa bundle exec rake db:setup?

Yes, that's why I run, I copied the wrong command.

So I rebuilt from scratch using my own fix and I got the same error. This is what I get for not trying my guesses... :sweat:  It actually seems like a bug in bundler: https://github.com/bundler/bundler/issues/5068.

Anyways, no need to block this just because of this tiny thing, totally fine with running an extra `bundle install` :)

**Comment by zuzust** [comment]
I cannot reproduce the problem as I have no `vendor/bundle` nor `.bundle/config` locally. Just a quick test: what happens if you add `.bundle` to .dockerignore?

**Comment by deivid-rodriguez** [comment]
I'm going to just remove those and try again. The only reason they are there is me trying to reproduce #30 locally. Hold on.

**Comment by deivid-rodriguez** [comment]
Nice, it worked! :tada: Let me run the rest of the setup, but this looks very promising :)

**Comment by deivid-rodriguez** [comment]
Confirming that the setup works for me. I could boot the app and visit the homepage at localhost:4000 and I could also run the tests. :+1:

**Comment by zuzust** [comment]
Ready to be merged! :clap:

File: .gitignore [hunk 1]
```diff
@@ -18,7 +18,10 @@
 # Ignore code coverage folder
 coverage/*
 
+config/docker/Dockerfile.*
```
Comment (deivid-rodriguez) [inline .gitignore]: One more question, what's this? Does docker generate a lock file or something like that?

Comment (zuzust) [inline .gitignore]: No, it's just that I prepared some variants of Dockerfile to test some pieces of the stack. May I remove that entry?

Comment (deivid-rodriguez) [inline .gitignore]: I see. I'd get rid of it then.

Comment (zuzust) [inline .gitignore]: Ok, working on it.

**Comment by zuzust** [comment]
`.gitignore` updated. All done!

**Comment by deivid-rodriguez** [comment]
Yeah, only thing to be aware of if/when this is to be deployed is to add the new config files with the right values beforehand. `mailserver.yml` is only used in dev, but the changes in redis config affect all environments.

**Comment by zuzust** [comment]
You're right. I will document it when the time comes.

File: config/docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+# PhantomJS
+ENV PHANTOMJS_VERSION 2.1.1
+RUN \
+  set -ex \
+  && wget -q -O phantomjs.tar.bz2 https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-${PHANTOMJS_VERSION}-linux-x86_64.tar.bz2 \
+  && mkdir -p /usr/local/share/phantomjs \
+  && tar -xjf phantomjs.tar.bz2 -C /usr/local/share/phantomjs --strip-components=1 \
+  && rm phantomjs.tar.bz2 \
+  && ln -s /usr/local/share/phantomjs/bin/phantomjs /usr/local/bin/phantomjs
```
Comment (deivid-rodriguez) [inline config/docker/Dockerfile]: `phantomjs 2.1.1` is available in `xenial`

```
$ apt policy phantomjs
phantomjs:
  Instalados: 2.1.1+dfsg-1
  Candidato:  2.1.1+dfsg-1
  Tabla de versión:
 *** 2.1.1+dfsg-1 500
        500 http://ubuntu.c3sl.ufpr.br/ubuntu xenial/universe amd64 Packages
        100 /var/lib/dpkg/status
```

So all that could be simplified, no? Or is there any reason to install it like that?

Comment (zuzust) [inline config/docker/Dockerfile]: Sh*t, I'll check it and refactor accordingly.

Comment (deivid-rodriguez) [inline config/docker/Dockerfile]: Right, then better to stick with the official binary. :+1:

**Pull request merged**